transport-protocol/src/custom.rs

```diff
@@ -23,35 +35,157 @@ pub enum TransportProtocolError {
+/// Return the length and the stripped buffer.
+/// This function must be called before deserializing a `protobuf` message
+fn get_length_prefix(buffer: &[u8]) -> (u64, &[u8]) {
+    let mut length_bytes: [u8; LENGTH_PREFIX_SIZE] = [0; LENGTH_PREFIX_SIZE];
+    length_bytes.copy_from_slice(&buffer[..LENGTH_PREFIX_SIZE]);
+    let remaining_buffer = &buffer[LENGTH_PREFIX_SIZE..];
+    (u64::from_be_bytes(length_bytes), remaining_buffer)
+}
+
+/// Return the input buffer prefixed with its length.
+/// This function must be called after serializing a `protobuf` message
+fn set_length_prefix(buffer: &mut Vec<u8>) -> TransportProtocolResult {
+    let length = u64::to_be_bytes(buffer.len() as u64);
+    let mut length_bytes = length.to_vec();
+    length_bytes.append(buffer);
+    Ok(length_bytes)
+}
+
+/// Strip the length prefix from the first chunk received and append subsequent
+/// chunks to the session's incoming buffer until the protocol buffer is
+/// complete and can be deserialized into a message.
+/// This function must be called everytime a new chunk is received.
+/// Take a session id and chunk as input.
+/// Return the complete buffer, or an error indicating that the buffer is
+/// partial, or any other error
+///
+/// TODO: harden this against potential malfeasance.  See the note, below.
+fn handle_protocol_buffer(session_id: Option<u32>, mut input: &[u8]) -> TransportProtocolResult {
+    // Default session id to 0
+    let session_id = session_id.unwrap_or(0);
+
+    let mut incoming_buffer_hash = INCOMING_BUFFER_HASH
+        .lock()
+        .map_err(|_| TransportProtocolError::MutexError(session_id))?;
+
+    // First, check if there is an entry in the hash for the specified session.
+    // If not, we assume this is the first chunk of the protocol buffer.
+    if incoming_buffer_hash.get(&session_id).is_none() {
+        if input.len() < LENGTH_PREFIX_SIZE {
+            return Err(TransportProtocolError::MissingLengthPrefix(session_id));
+        }
+
+        // Extract the protocol buffer's total length
+        let (expected_length, input_unprefixed) = get_length_prefix(&mut input);
+
+        // Insert the expected length in the hash table
+        incoming_buffer_hash.insert(session_id, (expected_length, Vec::new()));
+
+        input = input_unprefixed;
+    }
+
+    let (expected_length, incoming_buffer) = match incoming_buffer_hash.get_mut(&session_id) {
+        Some(v) => v,
+        None => return Err(TransportProtocolError::BufferNotFound(session_id)),
+    };
+
+    // Append chunk to incoming buffer
+    incoming_buffer.append(&mut input.to_vec());
+
+    // We return `Ok(None)` as long as the full protocol buffer has not yet been
+    // received. Once received, we return `Ok(buffer)`.
+    // In a well-behaving system, this is reasonable.  In a poorly-behaved
+    // system (under attack, clients just getting confused) it is not
+    // reasonable, and can eventually result in "Out of Memory" or Garbage out.
```

Comment:
Garbage out should never be possible, surely!

Reply:
Cf. comment above